vm/devices/storage/disklayer_ram/src/lib.rs

```diff
@@ -159,23 +198,25 @@ impl RamDiskLayer {
         sector: u64,
         overwrite: bool,
     ) -> Result<(), DiskError> {
-        let count = buffers.len() / SECTOR_SIZE as usize;
+        let count = buffers.len() >> self.sector_shift;
         tracing::trace!(sector, count, "write");
         let mut state = self.state.write();
         if sector + count as u64 > state.sector_count {
             return Err(DiskError::IllegalBlock);
         }
         for i in 0..count {
             let cur = i + sector as usize;
-            let buf = buffers.subrange(i * SECTOR_SIZE as usize, SECTOR_SIZE as usize);
+            let buf = buffers.subrange(i << self.sector_shift, 1 << self.sector_shift);
             let mut reader = buf.reader();
             match state.data.entry(cur as u64) {
                 Entry::Vacant(entry) => {
-                    entry.insert(Sector(reader.read_plain()?));
+                    let sector_data = reader.read_n(1 << self.sector_shift)?;
+                    entry.insert(Sector(sector_data));
                 }
                 Entry::Occupied(mut entry) => {
                     if overwrite {
-                        reader.read(&mut entry.get_mut().0)?;
+                        let sector_data = reader.read_n(entry.get_mut().0.len())?;
```

Comment:
Don't use read_n here, that reallocates unnecessarily. Read into the existing value.